[builder error] generic ptr (error: cannot convert 'struct main__ABCD' to 'struct main__ABCD *')

### Describe the bug

```v
module main

struct ObjectDesc {
	typ	u32
	ptr voidptr
}

struct ABCD {}

pub fn cast_object_desc[H](desc &ObjectDesc) ?H {
	$if H is &ABCD {
		if desc.typ == 12 { // desc == ABCD
			return &ABCD(desc.ptr)
		}
	}
	return none
}

fn main() {
	obj := ABCD{}
	cast_object_desc[&ABCD](ObjectDesc{ typ:12, ptr:voidptr(&obj) })
}
```

### Expected Behavior

readable error or ok

### Current Behavior

```
C:/Users/russo/AppData/Local/Temp/v_0/bug.14550208701212832316.tmp.c:7621: warning: cast between pointer and integer of different size
C:/Users/russo/AppData/Local/Temp/v_0/bug.14550208701212832316.tmp.c:12639: error: cannot convert 'struct main__ABCD' to 'struct main__ABCD *'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

`v bug.v`

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 fc4c431

### Environment details (OS name and version, etc.)

V full version: V 0.3.3 b2a71ec.fc4c431
OS: windows, Microsoft Windows 10 Pro v19042 64-bit
Processor: 16 cpus, 64bit, little endian,

getwd: C:\Users\russo\Desktop
vexe: D:\_dev\v_lang\v.exe
vexe mtime: 2023-04-04 20:52:46

vroot: OK, value: D:\_dev\v_lang
VMODULES: OK, value: C:\Users\russo\.vmodules
VTMP: OK, value: C:\Users\russo\AppData\Local\Temp\v_0

Git version: git version 2.35.3.windows.1
Git vroot status: weekly.2023.14-9-gfc4c431d
.git/config present: true

CC version: Error: exec failed (CreateProcess) with code 2: Не удается найти указанный файл.
 cmd: cc --version
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4755,7 +4755,7 @@ fn (mut g Gen) return_stmt(node ast.Return) {
 			styp := g.base_type(fn_ret_type)
 			g.writeln('${ret_typ} ${tmpvar};')
 			g.write('_option_ok(&(${styp}[]) { ')
-			if !fn_ret_type.is_ptr() && node.types[0].is_ptr() {
+			if !g.unwrap_generic(fn_ret_type).is_ptr() && node.types[0].is_ptr() {
 				if !(node.exprs[0] is ast.Ident && !g.is_amp) {
 					g.write('*')
 				}

--- a/vlib/v/tests/option_generic_ptr_return_test.v
+++ b/vlib/v/tests/option_generic_ptr_return_test.v
@@ -1,0 +1,22 @@
+struct ObjectDesc {
+	typ u32
+	ptr voidptr
+}
+
+struct ABCD {}
+
+pub fn cast_object_desc[H](desc &ObjectDesc) ?H {
+	$if H is &ABCD {
+		if desc.typ == 12 { // desc == ABCD
+			return &ABCD(desc.ptr)
+		}
+	}
+	return none
+}
+
+fn test_main() {
+	obj := ABCD{}
+	a := cast_object_desc[&ABCD](ObjectDesc{ typ: 12, ptr: voidptr(&obj) })
+	dump(a)
+	assert *a? == obj
+}

